### PR TITLE
Add HL7 protocol dissector

### PR DESCRIPTION
**Pull request by 0xA50C1A1**
Please sign (check) the below before submitting the Pull Request:

- [x] I have signed the ntop Contributor License Agreement at https://github.com/ntop/legal/blob/main/individual-contributor-licence-agreement.md
- [x] I have read the contributing guide lines at https://github.com/ntop/nDPI/blob/dev/CONTRIBUTING.md
- [x] I have updated the documentation (in doc/) to reflect the changes made (if applicable)

Describe changes:

Health Level 7 Version 2.x / 3 (over HTTP only) Messaging Standard - as you can guess from the name, it's widely used in medicine. I don't have any pcap samples of HL7 over HTTP, so I was only guided by [this description](https://v2plus.hl7.org/2021Jan/hl7-over-http.html).

### CODE REVIEW

**Comment by IvanNardi** [comment]
Is that https://github.com/virtalabs/tapirx/blob/develop/testdata/HL7-ORM.pcap the same HL7 protocol?

**Comment by 0xA50C1A1** [comment]
> Is that https://github.com/virtalabs/tapirx/blob/develop/testdata/HL7-ORM.pcap the same HL7 protocol?

Yea, it's HL7 version 2.x

File: src/lib/protocols/hl7.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    {
+      NDPI_LOG_INFO(ndpi_struct, "found HL7 over HTTP\n");
+      ndpi_set_detected_protocol(ndpi_struct, flow, NDPI_PROTOCOL_HL7, NDPI_PROTOCOL_HTTP, 
+                                 NDPI_CONFIDENCE_DPI);
+    }
+    return;
+  }
+
+  if (packet->payload_packet_len > 100 &&
+      ndpi_strnstr((const char *)packet->payload, "MSH|^~\\&|", packet->payload_packet_len))
```
Comment (IvanNardi) [inline src/lib/protocols/hl7.c]: Could we check for the magic string only at the beginning of the packet?

File: src/lib/protocols/hl7.c [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  if (((flow->detected_protocol_stack[0] == NDPI_PROTOCOL_HTTP) ||
+      (flow->detected_protocol_stack[1] == NDPI_PROTOCOL_HTTP)) &&
+      packet->content_line.ptr != NULL)
+  {
+    if ((LINE_ENDS(packet->content_line, "x-application/hl7-v2+er7") != 0) ||
+        (LINE_ENDS(packet->content_line, "x-application/hl7-v2+xml") != 0) ||
+        (LINE_ENDS(packet->content_line, "x-application/hl7-v3+xml") != 0) ||
+        (LINE_ENDS(packet->content_line, "x-application/fhir+xml") != 0)   ||
+        (LINE_ENDS(packet->content_line, "x-application/fhir+json") != 0)  ||
+        (LINE_ENDS(packet->content_line, "x-application/xml+cda") != 0))
```
Comment (IvanNardi) [inline src/lib/protocols/hl7.c]: Could you add this list in `fuzz/dictionary.dict` under a new section called HTTP Content Line, please?

File: src/lib/protocols/hl7.c [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+static void ndpi_search_hl7(struct ndpi_detection_module_struct *ndpi_struct,
+                            struct ndpi_flow_struct *flow)
+{
+  struct ndpi_packet_struct const * const packet = &ndpi_struct->packet;
+
+  NDPI_LOG_DBG(ndpi_struct, "search HL7\n");
+
+  if (((flow->detected_protocol_stack[0] == NDPI_PROTOCOL_HTTP) ||
+      (flow->detected_protocol_stack[1] == NDPI_PROTOCOL_HTTP)) &&
+      packet->content_line.ptr != NULL)
```
Comment (IvanNardi) [inline src/lib/protocols/hl7.c]: Two different ifs
```
  if (((flow->detected_protocol_stack[0] == NDPI_PROTOCOL_HTTP) ||
      (flow->detected_protocol_stack[1] == NDPI_PROTOCOL_HTTP)) {
    if(packet->content_line.ptr != NULL) {
       ...
    }
    return;
}
```
The goal is to return in any case if we are in the "HTTP-subprotocol" data path

**Review by IvanNardi — CHANGES_REQUESTED** [review]

**Comment by 0xA50C1A1** [comment]
@IvanNardi Thanks for the review, I've fixed all the issues.

**Review by IvanNardi — APPROVED** [review]
Thank you

**Pull request merged**